ksql-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectDataTranslator.java

```diff
@@ -0,0 +1,121 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+package io.confluent.ksql.serde.connect;
+
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.util.KsqlException;
+import org.apache.kafka.connect.data.Field;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.errors.DataException;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class ConnectDataTranslator {
+  public GenericRow toKsqlRow(final Schema schema, final Schema connectSchema,
+                              final Object connectData) {
+    if (! schema.type().equals(Schema.Type.STRUCT)) {
+      throw new KsqlException("Schema for a KSQL row should be a struct");
+    }
+    final Struct rowStruct = (Struct) toKsqlValue(schema, connectSchema, connectData);
+    return new GenericRow(
+        schema.fields()
+            .stream()
+            .map(f -> rowStruct.get(f.name()))
+            .collect(Collectors.toList())
+    );
+  }
+
+  private Object toKsqlValue(final Schema schema, final Schema connectSchema,
+                             final Object connectValue) {
+    // Map a connect value+schema onto the schema expected by KSQL. For now this involves:
+    // - handling case insensitivity for struct field names
+    // - setting missing values to null
+
+    if (!schema.type().equals(connectSchema.type())) {
+      throw new DataException(
+          String.format(
+              "Cannot deserialize type %s as type %s",
+              connectSchema.type().getName(),
+              schema.type().getName()));
+    }
+    if (connectValue == null) {
+      return null;
+    }
+    if (schema.type().isPrimitive()) {
+      return connectValue;
+    }
+    if (schema.type().equals(Schema.Type.ARRAY)) {
+      return toKsqlArray(schema.valueSchema(), connectSchema.valueSchema(), (List) connectValue);
+    }
+    if (schema.type().equals(Schema.Type.MAP)) {
+      return toKsqlMap(
+          schema.keySchema(), connectSchema.keySchema(),
+          schema.valueSchema(), connectSchema.valueSchema(), (Map) connectValue);
+    }
+    return toKsqlStruct(schema, connectSchema, (Struct) connectValue);
+  }
+
+  private List toKsqlArray(final Schema valueSchema, final Schema connectValueSchema,
+                           final List<Object> connectArray) {
+    return connectArray.stream()
+        .map(o -> toKsqlValue(valueSchema, connectValueSchema, o))
+        .collect(Collectors.toList());
+  }
+
+  private Map toKsqlMap(final Schema keySchema, final Schema connectKeySchema,
+                        final Schema valueSchema, final Schema connectValueSchema,
+                        final Map<String, Object> connectMap) {
+    return connectMap.entrySet().stream()
+        .collect(
+            Collectors.toMap(
+                e -> toKsqlValue(keySchema, connectKeySchema, e.getKey()),
```

Comment:
The map key is always `String`. Maybe we can enforce it here too.

Reply:
Its enforced by the schema translator. The data translator takes whatever schema ksql gives it and translates the deserialized value into that. So its up to the schema enforcement to ensure the stream/table schema uses string keys.

Orthogonally, this is actually another case worth discussing for schema translation. The connect translators handle non-string-keyed maps for us. Why keep that limitation in ksql?